### ISSUE TITLE
glibc version check problem

### ISSUE THREAD

**Issue by mlknz** [abseil/abseil-cpp#121]
Fails on `#if !__GLIBC_PREREQ(2, 12)` with newest android ndk.
Location of check: `abseil/absl/base/policy_checks.h:91`

Changing `#ifdef __GLIBC_PREREQ` for `#if defined(__GLIBC__) && defined(__GLIBC_PREREQ)` fixes problem.

Looked up solution there: ( https://github.com/openembedded/openembedded-core/blob/master/meta/recipes-extended/ltp/ltp/0008-Check-if-__GLIBC_PREREQ-is-defined-before-using-it.patch )

Some environment info:
- android ndk version `17.0.4754217`

- android sdk version `26`

- OS Ubuntu 16.04

- `ldd --version` outputs `(Ubuntu GLIBC 2.23-0ubuntu10) 2.23`

**Comment by derekmauro** [comment]
Can you post the exact error you are seeing? It looks like `__GLIBC_PREREQ` is defined but `__GLIBC__` isn't? That seems strange.

**Comment by mlknz** [comment]
```
[200/1509] Building CXX object %{ProjectDir}/libs/abseil/absl/types/bad_optional_access.cc.o
  FAILED: %{PathToAndroidSDK}/ndk-bundle/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang++  --target=armv7-none-linux-androideabi --gcc-toolchain=%{PathToAndroidSDK}/ndk-bundle/toolchains/arm-linux-androideabi-4.9/prebuilt/darwin-x86_64 --sysroot=%{PathToAndroidSDK}/ndk-bundle/sysroot  -DASIO_NO_EXCEPTIONS -DASIO_NO_TYPEID -DASIO_SEPARATE_COMPILATION -DASIO_STANDALONE -DFT2_BUILD_LIBRARY -DFT_CONFIG_OPTION_SYSTEM_ZLIB -DFT_CONFIG_OPTION_USE_BZIP2 -DGLM_FORCE_EXPLICIT_CTOR -DGOOGLE_PROTOBUF_NO_RTTI -DHAVE_BZIP2 -DHAVE_OPENSSL -DJSON_USE_EXCEPTION=0 -DMONOLITH -DOPENSSL_NO_ASM -DPNG_ARM_NEON_OPT=0 -DPUGIXML_NO_EXCEPTIONS -DPUGIXML_NO_XPATH -DZLIB -DZ_ENABLE_ASSERTIONS -DZ_ENABLE_INTROSPECTION -DZ_ENABLE_LOGGER -DZ_JNI_AVAILABLE -DZ_LOG_LEVEL=LogLevel::Debug -DZ_TARGET_ANDROID -D_USE_MATH_DEFINES
-isystem %{ManyLibs}
-I%{ProjectDir}/libs/abseil
-I%{MoreLibs}
-isystem %{PathToAndroidSDK}/ndk-bundle/sources/cxx-stl/llvm-libc++/include
-isystem %{PathToAndroidSDK}/ndk-bundle/sources/android/support/include
-isystem %{PathToAndroidSDK}/ndk-bundle/sources/cxx-stl/llvm-libc++abi/include
-isystem %{PathToAndroidSDK}/ndk-bundle/sysroot/usr/include/arm-linux-androideabi
-D__ANDROID_API__=17 -g -DANDROID -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -march=armv7-a -mfloat-abi=softfp -mfpu=vfpv3-d16 -mthumb -Wa,--noexecstack -Wformat -Werror=format-security -std=c++11  -fno-ident -fno-stack-protector -O0 -fno-limit-debug-info  -fPIC   -Wall -Wextra -Wpedantic -Wshadow -Wconversion -std=gnu++14  -frtti -fexceptions -MD -MT %{ProjectDir}/libs/abseil/absl/types/bad_optional_access.cc.o -MF %{ProjectDir}/libs/abseil/absl/types/bad_optional_access.cc.o.d -o %{ProjectDir}/libs/abseil/absl/types/bad_optional_access.cc.o -c %{ProjectDir}/libs/abseil/absl/types/bad_optional_access.cc
  In file included from %{ProjectDir}/libs/abseil/absl/types/bad_optional_access.cc:19:
  In file included from %{ProjectDir}/libs/abseil/absl/base/config.h:66:
 %{ProjectDir}/libs/abseil/absl/base/policy_checks.h:91:2: error: "Minimum required version of glibc is 2.12."
  #error "Minimum required version of glibc is 2.12."
```

Also important info: error wasn't happening with ndk version r15b.

**Comment by derekmauro** [comment]
I just checked the latest version of NDK and found this in `sources/cxx-stl/llvm-libc++/include/__config`

```
// Need to detect which libc we're using if we're on Linux.
#if defined(__linux__)
#include <features.h>
#if !defined(__GLIBC_PREREQ)
#define __GLIBC_PREREQ(a, b) 0
#endif // !defined(__GLIBC_PREREQ)
#endif // defined(__linux__)
```
so it looks like the change you suggested is necessary. I'll make this change for you. Thanks.

**Comment by derekmauro** [comment]
For anyone that finds this bug, it looks like LLVM decided that providing their own definition of `__GLIBC_PREREQ` was a bad idea and reverted it: https://github.com/llvm-mirror/libcxx/commit/fa614b847ac2f26a0ac5cdf83a558f52f88f054f

**Comment by derekmauro** [comment]
Fixed by add89fd0e4bfd7d874bb55b67f4e13bf8beca762.
